### PR TITLE
tsne plot integration

### PR DESCRIPTION
**Pull request by sdhossain**
allow for tsne plot integration. (Still not tested)

### CODE REVIEW

File: domainbed/lib/misc.py [hunk 1]
```diff
@@ -283,6 +289,62 @@ def accuracy(network, loader, weights, device, dataset):
...
+    ps = []
+    ys = []
+
+    network.eval()
+    with torch.no_grad():
+        i = 0
+        for x, y in loader:
+            x = x.to(device)
+            y = y.to(device)
+            z = network.featurizer(x) # SAGNet uses .network_f for featurizer
```
Comment (kimathikaai) [inline domainbed/lib/misc.py:306]: I'll need to validate that this is the best place to collect features for tsne plot generation

Comment (kimathikaai) [reply domainbed/lib/misc.py:306]:
```
if not isinstance(network, algorithms.SAGNet):
                z = network.featurizer(x) # SAGNet uses .network_f for featurizer
            else:
                z = network.network_f(x) # SAGNet uses .network_f for featurizer

```
Include this and import
`from domainbed import algorithms`

File: domainbed/lib/misc.py [hunk 2]
```diff
@@ -283,6 +289,62 @@ def accuracy(network, loader, weights, device, dataset):
...
+    df['domain'] = np.array([domain for _ in ys])
+
+    return df
+
+def get_tsne_plot(df):
+
+    # reduce dimensionality of featurized latent vectors
+    pca = PCA(n_components=48) 
+    pca.fit(df['latent_vector'])
+    df['pca_latent_vector']= pca.transform(source_zs[:n_samples])
```
Comment (kimathikaai) [inline domainbed/lib/misc.py:329]: `source_zs` and `n_samples` isn't defined

File: domainbed/scripts/train.py [hunk 3]
```diff
@@ -248,6 +255,12 @@ def save_checkpoint(filename):
                 results[name+'_f1'] = f1
                 results[name+'_nacc'] = non_overlap_class_acc
                 results[name+'_oacc'] = overlap_class_acc
+
```
Comment (kimathikaai) [inline domainbed/scripts/train.py:258]: Make the tsne plots per domain and include the domain in the name of the saved file (e.g. `tsne_plot_{domain}_{step}.png` but you can combine all the data in one csv file like you are doing now

**Review by kimathikaai — CHANGES_REQUESTED** [review]

File: domainbed/scripts/train.py (hunk 3 above)
Comment (sdhossain) [reply domainbed/scripts/train.py:258]: if we have a plot per domain, then wouldn't it be not possible to assess domain invariance?

**Pull request closed without merging**